### PR TITLE
Fixed #26028 -- Added an "Overriding Templates" howto.

### PR DESCRIPTION
**Pull request by phildini**
I think this could be accepted as-is, but I'm more looking for guidance than immediate merging. Is this the right place to put this? Is this too much detail or not enough?
https://code.djangoproject.com/ticket/26028

### CODE REVIEW

**Comment by carljm** [comment]
I think the content here is quite good; direct and clearly explained.

It seems like this type of content fits best (given how Django's docs are structured) within a "how-to" guide rather than in the topic introduction, which is already quite long. I'd suggest we add a new `howto/overriding-templates.txt`. That would also give us room to expand this a bit; I think it should at least briefly discuss the alternative approach of overriding one app from another app. That may not be what we want to present as the preferred approach, but there are some cases where it's useful, and if we're claiming to discuss "how to override templates" it should at least be mentioned (maybe with an example where one of your apps overrides one of its admin templates?).

I also think (as a somewhat separate but related change), the existing topic guide should somewhere clarify the order that apps are checked for templates (I believe it's in `INSTALLED_APPS` order, but that should be verified) when `APP_DIRS` is set to `True`. I'd add this near the top, under Configuration, in the bullet point where `APP_DIRS` is discussed.

File: docs/topics/templates.txt [hunk 1]
```diff
@@ -611,6 +611,49 @@ creating an object that specifies the following attributes:
...
+
+Overriding templates
+====================
+
+To override templates from a package in ``django.contrib`` or from a third-party
+package, the preferred method is to create replacement templates in your
+project's templates directory.
+
+Let's say you're trying to override the templates for a third-party package
+called ``blog``, which provides the templates ``post.html`` and ``list.html``.
```
Comment (carljm) [inline docs/topics/templates.txt]: For clarity, I think we should always refer to these templates as ``blog/post.html`` and ``blog/list.html``, never ``post.html`` and ``list.html``. It's clear from the directory structure you show later that that's the actual path to these templates. Since Django does _not_ do any automatic namespacing of template paths for templates within apps, and it's a common confusion point for newcomers who assume that it does, we should be careful not to encourage that mis-apprehension by giving unqualified paths to the templates.

File: docs/topics/templates.txt [hunk 2]
```diff
@@ -611,6 +611,49 @@ creating an object that specifies the following attributes:
...
+    INSTALLED_APPS = [
+        ...,
+        blog,
+        ...,
+    ]
+
+    TEMPLATES = [
+        {
+            'BACKEND': 'django.template.backends.django.DjangoTemplates',
+            'DIRS': ['templates'],
```
Comment (carljm) [inline docs/topics/templates.txt]: Does this actually work reliably? I thought you needed to do `os.path.join(BASE_DIR, 'templates')`

File: docs/topics/templates.txt [hunk 3]
```diff
@@ -611,6 +611,49 @@ creating an object that specifies the following attributes:
...
+
+    templates/
+        blog/
+            list.html
+            post.html
+
+When the configured template loader looks for templates for the ``blog`` app, it
+will look in the ``templates`` directory first, because that's what the settings
+above tell it to do. When the views in the ``blog`` app ask for the
+``post.html`` and ``list.html`` templates, the loader will return your project's
```
Comment (carljm) [inline docs/topics/templates.txt]: Use fully qualified paths again here.

File: docs/howto/overriding-templates.txt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        templates/
+            blog/
+                list.html
+                post.html
+
+With ``APP_DIRS`` set to ``True`` and the correct folder structure and templates,
+the template loader will look in the app's templates directory, find
+``blog/list.html`` and ``blog/post.html``, and return those.
+
+.. note::
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: Prefer `.. admonition:: Descriptive title goes here.` to generic "note".

File: docs/howto/overriding-templates.txt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    If you have a project templates dir, and an app templates dir, and they both
+    contain the overriden templates, the default Django template loader will
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: docs build failure: overridden

File: docs/howto/overriding-templates.txt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    TEMPLATES = [
+        {
+            ...,
+            'APP_DIRS': True,
+            ...
+        },
+    ]
+
+Similar to the example from the previous section, if the app we're working on is called ``myapp``, and the app we're trying to override is named ``blog``, then our directory
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: wrap lines at 79 chars

File: docs/howto/overriding-templates.txt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    TEMPLATES = [
+        {
+            'BACKEND': 'django.template.backends.django.DjangoTemplates',
+            'DIRS': [os.path.join(BASE_DIR,'templates')],
+            'APP_DIRS': True,
+            ...
+        },
+    ]
+
+The ``TEMPLATES`` section here should already exist in your project settings, as
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: Maybe linkify some of these settings? e.g.
```
:setting:`TEMPLATES`
```

File: docs/howto/index.txt [hunk 8]
```diff
@@ -28,6 +28,7 @@ you quickly accomplish common tasks.
    static-files/deployment
    windows
    writing-migrations
+   overriding-templates
```
Comment (timgraham) [inline docs/howto/index.txt]: please alphabetize this list

**Comment by timgraham** [comment]
Haven't taken a detailed look but there are a few high-level comments. Please remember that commit message should be in past tense.

**Comment by phildini** [comment]
I believe I have made all the changes suggested; let me know if there's anything else that we would like in this doc.

**Comment by berkerpeksag** [comment]
It would be nice to add a link to this guide in `docs/topics/templates.txt`.

The documentation only talks about overriding templates that use same backend `'django.template.backends.django.DjangoTemplates'`. It would be nice to mention about other backends too (I haven't used Jinja2 or any other backend yet so I don't know if that's possible).

File: docs/howto/overriding-templates.txt [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    INSTALLED_APPS = [
+        ...,
+        blog,
+        ...,
+    ]
+
+    TEMPLATES = [
+        {
+            'BACKEND': 'django.template.backends.django.DjangoTemplates',
+            'DIRS': [os.path.join(BASE_DIR,'templates')],
```
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: missing space after `BASE_DIR,`

File: docs/howto/overriding-templates.txt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+When the configured template loader looks for the ``blog`` app templates, it
+will look in the ``templates`` directory first, because that's what the settings
+above tell it to do. When the views in the ``blog`` app ask for the
+``blog/post.html`` and ``blog/list.html`` templates, the loader will return your
+project's copy of the templates.
+
+Overriding from an app's template directory
+===========================================
+
+While it is safer and cleaner to override templates from the project-level
```
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: I'd either explain why it's safer and cleaner or drop the first part of the sentence.

Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: I'd either explain why it's safer and cleaner or drop the first part of the sentence.

**Comment by berkerpeksag** [comment]
Could you add a short example of overriding a HTML file?

File: docs/howto/overriding-templates.txt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+templates directory, it is possible to override templates from within an app's
+template directory.
```
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: template -> templates

**Comment by berkerpeksag** [comment]
(Sorry for duplicate comments. GitHub is having some problems at the moment: https://status.github.com/messages)

File: docs/howto/overriding-templates.txt (hunk 10 above)
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: Slightly off-topic, but is there a section explains why it is safer to override templates from the project level templates directory?

**Comment by phildini** [comment]
I've made some of the suggested changes. The things that were mentioned above as "nice to have" I agree would be nice to have. Are they necessary for getting this landed or can we start with what's here?

Thanks for the reviews, happy to do what's necessary to get this landed!

File: docs/howto/overriding-templates.txt [hunk 12]
```diff
@@ -0,0 +1,88 @@
+====================
+Overriding Templates
+====================
+
+Overriding from the project's templates directory
+=================================================
+
+To override templates from a package in ``django.contrib`` or from a third-party
+package, the preferred method is to create replacement templates in your
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: preferred to who? Maybe something "A common way to override templates.... is to create replacement templates..." would be better.

File: docs/howto/overriding-templates.txt [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    INSTALLED_APPS = [
+        ...,
+        blog,
+        ...,
+    ]
+
+    TEMPLATES = [
+        {
+            'BACKEND': 'django.template.backends.django.DjangoTemplates',
+            'DIRS': [os.path.join(BASE_DIR, 'templates')],
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: Might want to add `BASE_DIR = ` with a note that it already exists in the default project template.

File: docs/howto/overriding-templates.txt [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            'APP_DIRS': True,
+            ...
+        },
+    ]
+
+The :setting:`TEMPLATES` section here should already exist in your project
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: section -> setting
should already exist -> will already exist if you created your project using the default project template.

File: docs/howto/overriding-templates.txt [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+settings, as it is normally created when you create a Django project. The
+setting that needs to be modified is :setting:`DIRS<TEMPLATES-DIRS>`.
+
+These settings assume you have a ``templates`` directory in the root of your
+project. To override the templates for the ``blog`` app, you'll create a folder
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: could chop "you'll"

File: docs/howto/overriding-templates.txt (hunk 10 above)
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: +1, I'd like to see this explained.

File: docs/howto/overriding-templates.txt [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+in the ``templates`` directory, and add the correct files to that folder::
+
+    templates/
+        blog/
+            list.html
+            post.html
+
+When the configured template loader looks for the ``blog`` app templates, it
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: This seems a bit wordy, perhaps something like "The template loader first looks for the blog app templates in the DIRS directory." would be enough.

File: docs/howto/overriding-templates.txt [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+will look in the ``templates`` directory first, because that's what the settings
+above tell it to do. When the views in the ``blog`` app ask for the
+``blog/post.html`` and ``blog/list.html`` templates, the loader will return your
+project's copy of the templates.
+
+Overriding from an app's template directory
+===========================================
+
+While it is safer and cleaner to override templates from the project-level
+templates directory, it is possible to override templates from within an app's
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: I'm in the habit of using contractions ("it's") or even "you can override" for a less stuffy tone.

File: docs/howto/overriding-templates.txt [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    TEMPLATES = [
+        {
+            ...,
+            'APP_DIRS': True,
+            ...
+        },
+    ]
+
+Similar to the example from the previous section, if the app we're working on is
+called ``myapp``, and the app we're trying to override is named ``blog``, then
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: chop first comma

File: docs/howto/overriding-templates.txt [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+                post.html
+
+With :setting:`APP_DIRS<TEMPLATES-APP_DIRS>` set to ``True`` and the correct
+folder structure and templates, the template loader will look in the app's
+templates directory, find ``blog/list.html`` and ``blog/post.html``, and return
+those.
+
+.. admonition:: Template loader ordering
+
+    If you have a project templates directory, and an app templates directory,
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: If you have app and project templates directories that both contain overridden templates, ...

File: docs/howto/overriding-templates.txt [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    and they both contain the overridden templates, the default Django template
+    loader will attempt to load the template from the project-level directory
+    first.
```
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: I think the point to make explicit is that DIRS and searched before APP_DIRS.

**Comment by timgraham** [comment]
After updating, don't forget to uncheck "Patch needs improvement" on the ticket so that it returns to the review queue.

**Comment by timgraham** [comment]
@phildini, do you have time to return to this?

**Comment by phildini** [comment]
I'm hoping to work on this soon; that said, if there's someone chomping at
the bit for it they can take it.

PJJ
http://philipjohnjames.com

File: docs/howto/overriding-templates.txt [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+
+With :setting:`APP_DIRS<TEMPLATES-APP_DIRS>` set to ``True`` and the correct
+folder structure and templates, the template loader will look in the app's
+templates directory, find ``blog/list.html`` and ``blog/post.html``, and return
+those.
+
+.. admonition:: Template loader ordering
+
+    If you have app and project templates directories that both contain
+    overrides , the default Django template loader will attempt to load the
```
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: Style nit: space before ,

File: docs/howto/overriding-templates.txt [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+    template from the project-level directory first. ``DIRS`` will be searched
```
Comment (berkerpeksag) [inline docs/howto/overriding-templates.txt]: This is a pretty minor thing, but you might want to add links to DIRS and APP_DIRS docs.

File: docs/howto/overriding-templates.txt (hunk 13 above)
Comment (timgraham) [inline docs/howto/overriding-templates.txt]: Did you address this comment?

File: docs/howto/index.txt [hunk 23]
```diff
@@ -17,15 +17,16 @@ you quickly accomplish common tasks.
...
    error-reporting
    initial-data
    jython
    legacy-databases
    outputting-csv
    outputting-pdf
+   overriding-templates
    static-files/index
    static-files/deployment
+   upgrade-version
```
Comment (timgraham) [inline docs/howto/index.txt]: I'd avoid doing this unrelated change in this PR.

**Comment by timgraham** [comment]
Edited and merged in 7c9a83330169df1118f6bd690aed131e7c59638d, thanks!
